### DISCUSSION TITLE
"Saving cache config" is always pretty long (with "TAM-verified archive" for every archive)

### DISCUSSION THREAD

**Discussion by szpak** [borgbackup/borg#7943] [Q&A] [answered]
I have ~10GB local repository with ~1,5K archives (backups) of multiple sets of data (multiple distinct source directories) done over a few years (kept in one repository to allow for deduplication on data hierarchy changes). All the directories are backed up together (one by one) on a regular basis. As number of changes in a particular source directories is usually small, the single "create" command takes <1s to process it, but after that there is ~30s delay on "Saving files cache". As that happens for every directory, the total backup time is significantly increased (in comparison to the actual "backup time" - without the post processing).

Sample command: `create --progress --compression auto,zstd --list --filter=AME --stats $BACKUP_PATH::$RESOURCE_NAME-'{hostname}-{now}' $PATH_TO_BACKUP`.

When executed with `--debug`, I clearly see that, during the aforementioned time, Borg (1.2.6 on Linux) always performs verification of the TAM authentication on every historical backup (~1,5K):

```
Archive authentication DISABLED.
TAM-verified archive
Archive authentication DISABLED.
TAM-verified archive
...
```

I performed the update procedure for 1.2.6 (although there might be some backups with it created with 1.2.6 due to the Borg update in the operating system) and there were no "tam:none", so according to the procedure, I haven't performed "borg upgrade". I'm not sure, if I did that on 1.0.9.

My related questions:
1. Should I see (as debug) "Archive authentication DISABLED" every time for my repository (somehow related to [that thread](https://github.com/borgbackup/borg/discussions/7895#discussioncomment-7411564))? Should that time confusing procedure be performed every time?
2. If yes, is it possible to explicitly disable that behavior on demand? I could just do it for the first directory from the set (and maybe for the last).
3. If yes, maybe there could some extra logging informing about "TAM authentication verification" or "Saving files cache" could be extended to mention also other postprocessing (I was initially thinking why saving cache for non changes takes so long)?

**Comment by ThomasWaldmann** [comment]
If you only have few changed source files, it is expected that the overhead of loading and saving the files cache, the chunks and repo index is more than the actual backup.

The archive TAM related stuff is relatively inexpensive as that is not much data to process, but of course the time needed for that scales with the archive count (btw, other borg operations, like `borg check` are impacted way more by that).

You can ignore the TAM related debug log messages, they are normal and only for debugging.

If your files cache is rather big and takes a lot of time to load/save **and** you are backing up different sets of source files using different `borg create` commands, you can split the files cache, see the environment variables documentation.

**Reply by szpak** [reply]
Thanks for the answer, but probably I wasn't precised enough.

Let's assume in one particular "borg create" operation, I have no changes in a directory with a few files with ~100MB in total (it's in a separate archive to be able to apply a different retention policy). The timing is as follows:
1. Checking for changed files - <1s
2. The "saving cache"-related logs - (Saving files cache, Saving chunks cache, Saving cache config) - <1s

```
Saving files cache
Saving chunks cache
Saving cache config
```

4. Integrity related logs - ~30s

```
Verified integrity of /my/repo/index.dddd
Archive authentication DISABLED.
TAM-verified archive
Archive authentication DISABLED.
TAM-verified archive
(two lines repeated ~1,5 times - the number of existing archives)
```

The borg cache itself is ~1GB:
```
$ du ~/.cache/borg/ -hs
1,1G	/home/foobar/.cache/borg/
```

The is also `Duration: 0.31 seconds` in the summary (which totally skips the post processing operations).

1. Is it really the cache issue to have ~30s of `TAM-verified archive` log entries? It seems to be ~1,5 times regardless what directory is archived (it seems that always all archives in a given repository are verified)?
2. If it is expected to have TAMs for all the archives verified every time, can it be disabled with a switch or with a env variable?

**Reply by ThomasWaldmann** [reply]
The repository manifest (which is loaded, modified and saved) contains a list of all archive infos (name, timestamp, archiveid).

Guess the code creates in-memory archive objects from that - either for some purpose or accidentally. For a simple `borg create` that should not be needed, but if you use `--stats`, maybe that triggers it.

Please edit your top post and give the `borg create` command you use (you can obfuscate the repo URL and the input files).

I'll have a look into the code afterwards...

**Reply by szpak** [reply]
Yes, I have the `--stats` switch there. I've updated the top post.

**Reply by ThomasWaldmann** [reply]
So, if you remove `--stats`, can you still observe the slow part?

**Reply by szpak** [reply]
It's definitely better. No extra delay. Everything within 2 seconds. Thanks for your suggestion!

However, I like `--stats` :-/. It allows me to generate a summary about increments in particular backups (by parsing the partial output logs). Do they need to always go together?

Update. Maybe I can generate stats for previous "create" execution(s)?

**Reply by ThomasWaldmann** [reply]
The best way to observe increments is to use `borg create --stats`.

Due to dedup, you can not compute that after the next backup was done.

But guess we can close this. "Everything comes at a price."

**Reply by szpak** [reply]
> But guess we can close this. "Everything comes at a price."

Good to have it explained :). Thanks.

**Reply by szpak** [reply, marked as answer]
As clarified by @ThomasWaldmann, it is not a problem with cache, but `--stats` generates extra time to go through all the existing archives in the repository which may take some time, if you have many of them (here 1,5K+).

**Reply by jdchristensen** [reply]
@ThomasWaldmann Is there any chance that something a bit inefficient is going on here?  For me, the part that reads the archives to generate the stats only takes about 4s (for 278 archives), which isn't terrible, but seems a bit long compared to how long other operations take.

**Reply by jdchristensen** [reply]
Or maybe there could be a variant of `--stats` which shows the `This archive:` line but not the `All archives:` line?  I'm not sure, but I'm guessing that it's only for the `All archives` line that all of the other archives need to be examined.

**Reply by ThomasWaldmann** [reply]
@jdchristensen your timings are quite similar to @szpak's timings, he has 6x as many archives as you and also about 6x the time needed for the stats.

Speed depends a bit on the borg version and the version of the archives - I remember doing some tuning / caching there in borg 1.2 (but first info/stats will be slow).

Stats related stuff will already be a bit different in borg2 (e.g. because I had to remove the csize values, because there we will always have some infos in the archives, etc.).

But some all archives or repo-wide stuff might be still slow - guess that would be an interesting experiment.
